orttraining/orttraining/core/graph/gradient_builder.cc

```diff
@@ -1683,41 +1682,68 @@ IMPLEMENT_GRADIENT_BUILDER(GetMinMaxGradient) {
 IMPLEMENT_GRADIENT_BUILDER(GetATenOpGradient) {
   std::vector<NodeDef> result;
   const auto& src_attrs = SrcNodeAttributes();
-  std::vector<AttributeProto> attrs;
   ORT_ENFORCE(utils::HasString(src_attrs.at("name")));
-  const std::string name = src_attrs.at("name").s();
-  attrs.emplace_back(MakeAttribute("name", name));
-  if (src_attrs.find("custom_attributes_json") != src_attrs.end()) {
-    attrs.emplace_back(MakeAttribute("custom_attributes_json", src_attrs.at("custom_attributes_json").s()));
-  }
-
-  const auto* op_config_ptr = contrib::aten_ops::ATenOperatorConfigs::Instance().GetConfig(name);
-  ORT_ENFORCE(op_config_ptr, "ATen Op config for ", name, " is not found.");
-  const auto& op_config = *op_config_ptr;
+  std::string op_name = src_attrs.at("name").s();
+  std::string overload_name = "";
+  if (src_attrs.find("overload_name") != src_attrs.end() && utils::HasString(src_attrs.at("overload_name"))) {
+    overload_name = src_attrs.at("overload_name").s();
+  }
+
+  const auto& grad_def = ATenOpGradientDefinitionGetter::Instance()(op_name, overload_name);
+  ORT_ENFORCE(!grad_def.empty());
+
+  std::unordered_set<std::string> seen_outputs;
+  for (const auto& node_def : grad_def) {
+    OpDef op_def(node_def.op_type, node_def.domain);
+    bool is_aten_op = node_def.op_type == "ATenOp" && node_def.domain == kMSDomain;
+    std::vector<ArgDef> input_args;
+    for (const auto& input : node_def.inputs) {
+      if (input.find("GO(") == 0) {
+        int index = std::stoi(input.substr(3, input.length() - 4));
+        input_args.emplace_back(GO(static_cast<size_t>(index)));
+      } else if (input.find("I(") == 0) {
+        int index = std::stoi(input.substr(2, input.length() - 3));
+        input_args.emplace_back(I(static_cast<size_t>(index)));
+      } else if (input.find("O(") == 0) {
+        int index = std::stoi(input.substr(2, input.length() - 3));
+        input_args.emplace_back(O(static_cast<size_t>(index)));
+      } else {
+        ORT_ENFORCE(seen_outputs.find(input) != seen_outputs.end(), input, " is not a valid intermediate output.");
+        input_args.emplace_back(IA(input));
+      }
+    }
 
-  std::vector<int64_t> grad_output_types;
-  std::vector<ArgDef> input_args;
-  std::vector<ArgDef> output_args;
+    std::vector<ArgDef> output_args;
+    std::vector<int64_t> grad_output_types;
+    for (const auto& output : node_def.outputs) {
+      if (output.find("GI(") == 0) {
+        size_t index = static_cast<size_t>(std::stoi(output.substr(3, output.length() - 4)));
+        output_args.emplace_back(GI(index));
+        if (is_aten_op) {
+          grad_output_types.emplace_back(IElemType(index));
```

Comment:
in addition to type, we can also pass shape here.

Reply:
List of list (may have different sizes) is not easy to support in node attribute...